src/python_framework_v02/nhd_network.py

```diff
@@ -2,17 +2,20 @@
 from itertools import chain
 from functools import reduce, partial
 from collections.abc import Iterable
+import logging
 
+LOG = logging.getLogger(__name__)
 
 def nodes(N):
     yield from N.keys() | (v for v in chain.from_iterable(N.values()) if v not in N)
+    LOG.debug("Edges: %s", N)
```

Comment:
Note that N can be be quite large. Printing the network might not be a very useful log message. A perhaps, more useful log message might be the expected number of nodes in N `len(N) + sum(map(len, N.values()))`